File: web/Vector.js
class Vector {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    add(vec) {
        this.x += vec.x;
        this.y += vec.y;
        return this;
    }


    sub(vec) {
        this.x -= vec.x;
        this.y -= vec.y;
        return this;
    }


    normalize() {
        let norm = this.magnitude();
        this.x /= norm;
        this.y /= norm;
        return this;
    }


    magnitude() {
        return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2));
    }

    mult(n) {
        this.x *= n;
        this.y *= n;
        return this;
    }


    div(n) {
        this.x /= n;
        this.y /= n;
        return this;
    }

    inverse() {
        this.x -= 1;
        this.y -= 1;
        return this;
    }


    copy() {
        return Object.assign(new Vector(0, 0), this);
    }

    static div(vec1, n) {
        return vec1.copy().div(n);
    }

    static mult(vec1, n) {
        return vec1.copy().mult(n);
    }
    static add(vec1, vec2) {
        return vec1.copy().add(vec2);
    }
    static sub(vec1, vec2) {
        return vec1.copy().sub(vec2);
    }
    static normalize(vec) {
        return vec.copy().normalize();
    }
    static inverse(vec1) {
        return vec1.copy().inverse();
    }

    static dot(vec1, vec2) {
        return vec1.x * vec2.x + vec1.y * vec2.y;
    }

    static dist(vec1, vec2) {
        return Vector.sub(vec1, vec2).magnitude();
    }
    static get Unit() {
        return {
            up: new Vector(0, -1),
            upleft: new Vector(-Math.SQRT1_2, -Math.SQRT1_2),
            left: new Vector(-1, 0),
            downleft: new Vector(-Math.SQRT1_2, Math.SQRT1_2),
            down: new Vector(0, 1),
            downright: new Vector(Math.SQRT1_2, Math.SQRT1_2),
            right: new Vector(1, 0),
            upright: new Vector(Math.SQRT1_2, -Math.SQRT1_2),
        };
    }
}

File: pyrl/Vector.py
import math
from enum import Enum

class Vector:
    
    def __init__(self, x, y):
    
        self.x = x
        self.y = y
        
    def add(self, vec2):
        result = Vector(self.x + vec2.x, self.y + vec2.y)
        return result
    

    def sub(self, vec2):
        result = Vector(self.x - vec2.x, self.y - vec2.y)
        return result
      
    def mult(self, n):
        result = Vector(self.x * n, self.y * n)
        return result
    
    def div(self, n):
        result = Vector(self.x / n, self.y / n)
        return result
        
    def magnitude(self):
        return (self.x ** 2 + self.y ** 2) ** 0.5 
        
    def normalize(self):
        norm = self.magnitude()
        result = Vector(self.x / norm, self.y / norm)
        return result

    def inverse(self):
        result = self.mult(-1)
        return result
    
    def copy(self):
        return Vector(self.x, self.y)

    def dot(self, vec2):
        result = self.x * vec2.x + self.y * vec2.y
        return result
    
    def dist(self, vec2):
        return Vector.sub(self, vec2).magnitude()
        
    def __repr__(self):
        return "Vector: ({}, {})".format(self.x, self.y)
    
    def __str__(self):
        return "Vector: ({}, {})".format(self.x, self.y)
    
    def __eq__(self, vec2):
        return self.x == vec2.x and self.y == vec2.y
    

class UnitVec:
    up = Vector(0, -1)
    upleft = Vector(-math.sqrt(0.5), -math.sqrt(0.5))
    left = Vector(-1, 0)
    downleft = Vector(-math.sqrt(0.5), math.sqrt(0.5))
    down = Vector(0, 1)
    downright = Vector(math.sqrt(0.5), math.sqrt(0.5))
    right = Vector(1, 0)
    upright = Vector(math.sqrt(0.5), -math.sqrt(0.5))
